protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,638 @@
+    }
+
+    fn handle_response(
+        &mut self,
+        sender: PeerId,
+        response: Option<DialResponse>,
+    ) -> Option<NatStatus> {
+        let mut probe = self.ongoing_outbound.take()?;
+        if !probe.pending_servers.contains(&sender) {
+            let _ = self.ongoing_outbound.insert(probe);
+            return None;
+        };
+        probe.pending_servers.retain(|p| p != &sender);
+        match response {
+            Some(DialResponse::Ok(addr)) => self.handle_dial_ok(probe, addr),
+            Some(DialResponse::Err(err)) => {
+                if !matches!(err, ResponseError::DialError) {
+                    probe.dismissed += 1;
+                }
+                probe.errors.push((sender, err));
+                self.handle_dial_err(probe)
+            }
+            None => {
+                probe.dismissed += 1;
+                self.handle_dial_err(probe)
+            }
+        }
+    }
+
+    fn handle_dial_ok(&mut self, mut probe: Probe, address: Multiaddr) -> Option<NatStatus> {
+        let score = probe.addresses.get_mut(&address)?;
+        *score += 1;
+        if *score < probe.required_success {
+            let _ = self.ongoing_outbound.insert(probe);
+            return None;
+        }
+        Some(NatStatus::Public {
+            address: address.clone(),
+            server_count: probe.server_count,
+            successes: *score,
+        })
+    }
+
+    fn handle_dial_err(&mut self, probe: Probe) -> Option<NatStatus> {
+        let remaining = probe.pending_servers.len();
+        let current_highest = probe.addresses.iter().max_by(|(_, a), (_, b)| a.cmp(b))?.1;
+
+        // Check if the probe can still be successful.
+        if remaining >= probe.required_success - current_highest {
+            let _ = self.ongoing_outbound.insert(probe);
+            return None;
+        }
+
+        let valid_attempts = probe.server_count - remaining - probe.dismissed;
+        if valid_attempts >= probe.min_valid {
+            Some(NatStatus::Private {
+                server_count: probe.server_count,
+                errors: probe.errors,
+                tried_addresses: probe.addresses.into_iter().collect(),
+            })
+        } else {
+            Some(NatStatus::Unknown {
+                server_count: probe.server_count,
+                errors: probe.errors,
+                tried_addresses: probe.addresses.into_iter().collect(),
+            })
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ProtocolsHandler = <RequestResponse<AutoNatCodec> as NetworkBehaviour>::ProtocolsHandler;
+    type OutEvent = NatStatus;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        self.inner.new_handler()
+    }
+
+    fn addresses_of_peer(&mut self, peer: &PeerId) -> Vec<Multiaddr> {
+        if let Some(addrs) = self.ongoing_inbound.get(peer).map(|(a, _)| a.clone()) {
+            addrs
+        } else {
+            self.inner.addresses_of_peer(peer)
```

Comment:
If I am not mistaken `RequestResponse::add_address` is never called. With that in mind, is the delegation to `self.inner` needed here?

Reply:
`RequestResponse::addresses_of_peer` also includes the addresses of the currently connected peers, which we need if we do a probe with `extend_with_connected = true`.
Apart from that, with https://github.com/libp2p/rust-libp2p/pull/2262/commits/0a612c07cd1d7f985f4e20b44a8817bcb57defac now also the addresses of the "static servers" are added via the inner request-response protocol.